pkg/storegateway/indexcache/cache.go

```diff
@@ -116,6 +118,22 @@ type IndexCache interface {
 	FetchLabelValues(ctx context.Context, userID string, blockID ulid.ULID, labelName string, matchersKey LabelMatchersKey) ([]byte, bool)
 }
 
+// BlockTTL determines an appropriate TTL for cache entries related to a particular block
+// based on metadata about the block. Use a shorter TTL for temporary blocks that will be
+// compacted and deleted soon and a longer TTL for larger blocks that won't be deleted.
+func BlockTTL(meta *block.Meta) time.Duration {
+	duration := time.Duration(meta.MaxTime-meta.MinTime) * time.Millisecond
+
+	// Anything less than or equal to 12h is a temporary block that will eventually
+	// be compacted into a 24h block. Don't cache for long since otherwise these will
+	// stay in the cache long after they're no longer relevant.
+	if duration <= 12*time.Hour {
+		return 2 * time.Hour
```

Comment:
Why a 12h block is cached for 2h? It will be queried for longer than 12h. Realistically it will take at least 12h before a **new** 12h block will be compated into a 24h block. Have you considered simply returning `duration`?

Reply:
Sure I can do that. My main goal was just to use something less than 7 days and there was a lot more variation in block durations than I expected. 

For example, a snapshot from an internal cluster yesterday, the space taken in and index cache instance grouped by the duration of the associated block (rounded to nearest 1h):

```
24h0m0s 892388444
22h0m0s 2259095
20h0m0s 2834155
14h0m0s 3167
12h0m0s 46513828
10h0m0s 877425
8h0m0s  147618
6h0m0s  20840
2h0m0s  111656
0s      1177989
```

I'll make the change to use `duration` when it's less than 24h and 7d when it's equal or larger.